### ISSUE TITLE
[Question] How to serialize sum type to json?

### ISSUE THREAD

**Issue by hh9527** [tweag/nickel#1993]
```nickel
let Ty = [| 'Num Number, 'Str String |] in 'Num 1 | Ty
```
```text
$ nickel export t.ncl
error: non serializable term
```

**Comment by yannham** [comment]
Indeed, enums (at least with arguments) can't currently be serialized. Quoting [the manual](https://nickel-lang.org/user-manual/syntax#enum-variants):

> Warning 2: enum variants are the only primitive data structure of Nickel that can't be serialized. Indeed, there is no obvious canonical way to encode enum variants in the JSON data model (though many such encodings exist). If you need to serialize and deserialize enum variants, you'll have to explicitly map them to and from serializable data structures (such as records). In general, enum variants are rather expected to be used internally to make nice and ergonomic library APIs.

In particular, if you just want to store a value that is either a string or a number, and serialize it without any tag (that is serialize `'Num 1` as `1` and `'Str "a"` as `"a"`), you could probably directly store either a `Number` or a `String` in a field directly. You can write a contract `NumberOrString` for that. In the coming new version of Nickel, you'll even have an operator (the name is not fixed yet) to do that:

```nickel
> let Ty = std.contract.any_of [Number, String] in 1 | Ty
> 1 | Ty
1

> "a" | Ty
"a"
```

If for any reason you really need to keep the enum, you have to transform it to the underlying value at some point. Something like https://github.com/tweag/nickel/issues/1170#issuecomment-1461637685 could also help, but it's an idea and isn't yet implemented, unfortunately.

**Comment by hh9527** [comment]
@yannham Thank you for your explaination.

You explaination is very clear, I think I had use nickel in the wrong way.

nickel assume the users to use it in the data-centric-way, but I use it in the type-centric-way.

**Comment by yannham** [comment]
You're welcome! I'm going to close for now as the original question was mostly asking about information. Feel free to open other issues if there are other questions or proposals around enums and enums serialization in particularly.

**Comment by rben01** [comment]
I also faced this issue, and this GitHub issue was the only solution I found. I think that in the docs, in the enum section, there should be a note about the pitfalls of enums (no serializability) and an explanation of how to reimplement them using `std.contract.any_of` — especially since it seems like `std.contract.any_of` is the proper user-facing tool (i.e. you should vend `std.contract.any_of` types/contracts, not enums, for users of your schemas).
